src/Arcus.Messaging.Tests.Integration/MessagePump/ServiceBusMessagePumpTests.cs

```diff
@@ -280,11 +281,53 @@ await using (var service = await TestMessagePumpService.StartNewAsync(config, _l
             }
         }
 
+        [Fact]
+        public async Task ServiceBusMessagePump_RotateServiceBusConnectionKeysOnSecretNewVersionNotification_MessagePumpRestartsThenMessageSuccessfullyProcessed()
+        {
+            // Arrange
+            var config = TestConfig.Create();
+            KeyRotationConfig keyRotationConfig = config.GetKeyRotationConfig();
+            _logger.LogInformation("Using Service Principal [ClientID: '{0}']", keyRotationConfig.ServicePrincipal.ClientId);
+
+            var client = new ServiceBusConfiguration(keyRotationConfig, _logger);
+            string freshConnectionString = await client.RotateConnectionStringKeysForQueueAsync(KeyType.PrimaryKey);
+
+            ServicePrincipalAuthentication authentication = keyRotationConfig.ServicePrincipal.CreateAuthentication();
+            IKeyVaultClient keyVaultClient = await authentication.AuthenticateAsync();
+            await SetConnectionStringInKeyVaultAsync(keyVaultClient, keyRotationConfig, freshConnectionString);
+
+            var commandArguments = new[]
+            {
+                CommandArgument.CreateSecret("EVENTGRID_TOPIC_URI", config.GetTestInfraEventGridTopicUri()),
+                CommandArgument.CreateSecret("EVENTGRID_AUTH_KEY", config.GetTestInfraEventGridAuthKey()),
+                CommandArgument.CreateSecret("ARCUS_KEYVAULT_VAULTURI", keyRotationConfig.KeyVault.VaultUri),
+                CommandArgument.CreateSecret("ARCUS_KEYVAULT_CONNECTIONSTRINGSECRETNAME", keyRotationConfig.KeyVault.SecretName),
+                CommandArgument.CreateSecret("ARCUS_KEYVAULT_SERVICEPRINCIPAL_CLIENTID", keyRotationConfig.ServicePrincipal.ClientId),
+                CommandArgument.CreateSecret("ARCUS_KEYVAULT_SERVICEPRINCIPAL_CLIENTSECRET", keyRotationConfig.ServicePrincipal.ClientSecret),
+                CommandArgument.CreateSecret("ARCUS_KEYVAULT_SECRETNEWVERSIONCREATED_CONNECTIONSTRING", keyRotationConfig.KeyVault.SecretNewVersionCreated.ConnectionString)
+            };
+
+            using (var project = await WorkerProject.StartNewWithAsync<ServiceBusQueueSecretNewVersionReAuthenticateProgram>(config, _logger, commandArguments))
+            {
+                string newSecondaryConnectionString = await client.RotateConnectionStringKeysForQueueAsync(KeyType.SecondaryKey);
```

Comment:
What if we run await service.SimulateMessageProcessingAsync before rotating it? That makes sure it is first using the old connection string and works fine.

After that, we rotate the connection and run it again which should be fine as well.

Reply:
Hmm, isn't that what the other tests are for? It's some while ago when I wrote this, but I think I chose to rotate first because that would make sure that I got a connection string that's still valid.
I guess this test is more about restarting and the final check about simulating a message is the assertion-proof that after all the key rotations and restarts we can still receive messages.

We rotate 3 times here in the test. Something we agreed on was a concrete example of how an application would rotate it.